README.md
# libpuppetdb

## Intro

libpuppetdb is a simple library for the execution of Puppet DB queries.
The library uses libcurl and relies on the Puppet DB REST API.
Both both SSL (HTTP) and non-SSL (HTTPS) queries are supported.

## Using libpuppetdb

To use libpuppetdb in your code, you first create a PuppetdbConnector object.
You do that by specifying the Puppet DB host and, for secure connections, the
SSL certificates. Then you execute queries by instantiating Query objects and
calling PuppetdbConnector::performQuery, that returns the query results
(json format) as a string.

## Errors

lippuppetdb errors follow a class hierarchy where the parent class is
`libpuppetdb_error`. Specific errors are:

| name | description
|------|------|------------
| query_error | thrown by the Query constructor
| connector_error | thrown by the PuppetdbConnector constructor
| processing_error | thrown by the PuppetdbConnector::performQuery method

Note that `processing_error` objects contain an indication of possible
libcurl errors.

## Create a connection

To create a connection, you have to specify the Puppet DB host name. The service
port (default 8080) and the Puppet DB API version (default 'v3') are optional
arguments.
To create a secure connection, you also need to provide the file paths of the
required certificates, as descrbied in the Puppet DB online documenation:
the CA certificate, the client SSL certificate, and the client private key.

## Create a query

To create a query, you simply specify the Puppet DB endpoint (mandatory) and
the query string (optional).

## Build Requirements

libpuppetdb is written in C++11, so you must call your compiler accordingly.

The libcurl dev library is required. It must be SSL-enabled for secure
connections.

## Testing

Inside the test directory, run 1) "cmake .", 2) "make", and 3) "make test".
A simple non-SSL example is provided in the example directory.


CHANGELOG.md
# Changelog

Change history for lippuppetdb

## 0.2.0

Released 2015/05/15

* Implementing exception handling

## 0.1.0

Released 2015/05/01

* Initial release, supporting the puppetdb api v4 (currently experimental)


test/travis_target.sh
#! /bin/bash

$HOME/bin/cmake .
if [ $? -ne 0 ]; then
    echo "cmake failed."
    exit 1
fi

make
if [ $? -ne 0 ]; then
    echo "test building failed."
    exit 1
fi

make test


test/libpuppetdb_test.cpp
/*
    libpuppetdb_test.cpp
    ======================

    libpuppetdb unit tests.
    Requires: googletest, googlemock.

*/

#include "../include/libpuppetdb/libpuppetdb.h"
#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include <chrono>

// To test SSL with puppet agent certificates
// #define PUPET_AGENT_TEST

namespace LibPuppetdb {

// Mock classes

class MockURLConnector : public PuppetdbConnector {
  public:
    MockURLConnector(std::string host_name)
        : PuppetdbConnector::PuppetdbConnector(host_name) {}
    MOCK_METHOD2(getQueryUrl, std::string(Query& query, CURL* curl));
};


class MockSetupConnector : public PuppetdbConnector {
  public:
    MockSetupConnector(std::string host_name)
        : PuppetdbConnector::PuppetdbConnector(host_name) {}
    MOCK_METHOD1(setupAndPerform, std::string(Query& query));
};

// Testing the PuppetDB query

class QueryTest : public ::testing::Test {};

TEST_F(QueryTest, validQuery) {
    EXPECT_NO_THROW(Query("facter"));
}

TEST_F(QueryTest, validQueryWithQueryString) {
    EXPECT_NO_THROW(Query("nodes", "puppetdb_query"));
}

TEST_F(QueryTest, invalidQuery) {
    EXPECT_THROW(Query(""), query_error);
}

TEST_F(QueryTest, invalidQueryWithQueryString) {
    EXPECT_THROW(Query("", "puppetdb_query"), query_error);
}

TEST_F(QueryTest, getQueryComponents) {
    std::string endpoint { "eggs" };
    Query query { endpoint };

    EXPECT_EQ(endpoint, query.getEndpoint());
    EXPECT_EQ("", query.getQueryString());
}

TEST_F(QueryTest, getQueryComponentsWithQueryString) {
    std::string endpoint { "foo" };
    std::string query_string { "bar" };
    std::string expected_query { endpoint + "?query=" + query_string };
    Query query { endpoint, query_string };

    std::string obtained_query { query.getEndpoint() + "?query="
                                 + query.getQueryString() };
    EXPECT_EQ(expected_query, obtained_query);
}

// Testing the PuppetDB connector

class ConnectionTest : public ::testing::Test {
  protected:
    CURL* curl_handle_;

    virtual void SetUp() {
        curl_handle_ = curl_easy_init();
    }

    virtual void TearDown() {
        curl_easy_cleanup(curl_handle_);
    }
};

TEST_F(ConnectionTest, ValidConnector) {
    EXPECT_NO_THROW(PuppetdbConnector("eggs"));
}

TEST_F(ConnectionTest, ValidConnectorWithoutSSL) {
    EXPECT_NO_THROW(PuppetdbConnector("eggs", 42, ApiVersion::v3));
}

TEST_F(ConnectionTest, isSecure) {
    PuppetdbConnector connector { "spam", 42, ApiVersion::v2 };

    EXPECT_FALSE(connector.isSecure());
}

TEST_F(ConnectionTest, ConnectDefaultPortAndVersion) {
    PuppetdbConnector connector { "spam" };
    Query query { "facts" };
    std::string expected_url { "http://spam:8080/v4/facts" };

    EXPECT_FALSE(connector.isSecure());
    EXPECT_EQ(expected_url, connector.getQueryUrl(query, curl_handle_));
}

TEST_F(ConnectionTest, ConnectWithoutHost) {
    EXPECT_THROW(PuppetdbConnector(""), connector_error);
}

TEST_F(ConnectionTest, ConnectWithSSLNonExistentCerts) {
    auto ctor = []() { PuppetdbConnector("fake_host",
                                         "/fake/path/ca.cer",
                                         "/fake/path/host.cer",
                                         "/fake/path/host.key");
                     };

    EXPECT_THROW(ctor(), connector_error);
}

TEST_F(ConnectionTest, ConnectWithSSLValid) {
    auto ctor = []() { PuppetdbConnector("fake_host",
                                         "./resources/ca_crt.pem",
                                         "./resources/test_crt.pem",
                                         "./resources/test_key.pem");
                     };

    EXPECT_NO_THROW(ctor());
}

TEST_F(ConnectionTest, ConnectWithSSLIsSecure) {
    PuppetdbConnector connector { "fake_host",
                                  "./resources/ca_crt.pem",
                                  "./resources/test_crt.pem",
                                  "./resources/test_key.pem" };

    EXPECT_TRUE(connector.isSecure());
}

TEST_F(ConnectionTest, performQueryWithSimpleResult) {
    MockSetupConnector mock_connector { "bar" };
    Query query { "foo" };

    EXPECT_CALL(mock_connector, setupAndPerform(testing::_))
        .WillOnce(testing::Return("simple_result"));

    std::string result { mock_connector.performQuery(query) };

    EXPECT_EQ("simple_result", result);
}

TEST_F(ConnectionTest, httpQuery) {
    MockURLConnector mock_connector { "bar" };
    Query query { "foo" };
    EXPECT_CALL(mock_connector, getQueryUrl(testing::_, testing::_))
        .WillOnce(testing::Return("http://example.com"));

    std::string result { mock_connector.performQuery(query) };

    EXPECT_FALSE(result.size() == 0);
}

TEST_F(ConnectionTest, multipleHttpQueries) {
    MockURLConnector mock_connector { "spam" };
    Query first_query { "eggs" };
    Query second_query { "beans" };
    EXPECT_CALL(mock_connector, getQueryUrl(testing::_, testing::_))
        .WillOnce(testing::Return("http://example.com"))
        .WillOnce(testing::Return("http://goolge.com"));

    std::string first_result { mock_connector.performQuery(first_query) };
    std::string second_result { mock_connector.performQuery(second_query) };

    EXPECT_FALSE(first_result.size() == 0);
    EXPECT_FALSE(second_result.size() == 0);
}

#ifdef PUPET_AGENT_TEST
    TEST_F(ConnectionTest, testingPuppetSSL) {
        PuppetdbConnector connector {
            "master",
            "/etc/puppetlabs/puppet/ssl/certs/ca.pem",
            "/etc/puppetlabs/puppet/ssl/certs/centos65a.pem",
            "/etc/puppetlabs/puppet/ssl/private_keys/centos65a.pem" };
        Query query { "facts" };
        std::string result = connector.performQuery(query);
        std::cout << "Facter result:\n" << result << "\n\n";

        EXPECT_FALSE(result.size() == 0);
    }
#endif

}  // namespace LibPuppetdb


test/puppetdbquery_test.cpp
/*
    puppetdbquery_test.cpp
    ======================

    PuppetDB Query unit tests.
    Requires: googletest, googlemock.

*/


#ifndef PUPPETDBQUERY_TEST_PUPPETDB_QUERY_TEST_H_
#define PUPPETDBQUERY_TEST_PUPPETDB_QUERY_TEST_H_

#include "../include/puppetdbquery/puppetdbquery.h"
#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include <chrono>

// To test SSL with puppet agent certificates
// #define PUPET_AGENT_TEST

namespace PuppetdbQuery {


// Mock classes

class MockURLConnector : public PuppetdbConnector {
  public:
    using PuppetdbConnector::PuppetdbConnector;
    MOCK_METHOD2(getQueryUrl, std::string(Query& query, CURL* curl));
};


class MockSetupConnector : public PuppetdbConnector {
  public:
    using PuppetdbConnector::PuppetdbConnector;
    MOCK_METHOD1(setupAndPerform, std::string(Query& query));
};



// Testing the PuppetDB query

class QueryTest : public ::testing::Test {};


TEST_F(QueryTest, validQuery) {
    Query query {"facter"};
    Query query_with_query {"nodes", "puppetdb_query"};

    EXPECT_TRUE(query.isValid());
    EXPECT_EQ(static_cast<int>(ErrorCode::OK), query.getErrorCode());
    EXPECT_TRUE(query_with_query.isValid());
    EXPECT_EQ(static_cast<int>(ErrorCode::OK), query_with_query.getErrorCode());
}

TEST_F(QueryTest, invalidQuery) {
    Query query {""};
    Query query_with_query {"", "puppetdb_query"};

    EXPECT_FALSE(query.isValid());
    EXPECT_EQ(static_cast<int>(ErrorCode::INVALID_QUERY), query.getErrorCode());
    EXPECT_FALSE(query_with_query.isValid());
    EXPECT_EQ(static_cast<int>(ErrorCode::INVALID_QUERY),
              query_with_query.getErrorCode());
}

TEST_F(QueryTest, setAndGetErrorCode) {
    Query query {"spam"};
    int error_code {42};
    query.setErrorCode(error_code);

    EXPECT_EQ(error_code, query.getErrorCode());
}

TEST_F(QueryTest, getQueryComponents) {
    std::string endpoint {"eggs"};
    Query query {endpoint};

    EXPECT_EQ(endpoint, query.getEndpoint());
    EXPECT_EQ("", query.getQueryString());
}

TEST_F(QueryTest, getQueryComponentsWithQueryString) {
    std::string endpoint {"foo"};
    std::string query_string {"bar"};
    std::string expected_query {endpoint + "?query=" + query_string};
    Query query {endpoint, query_string};

    std::string obtained_query {query.getEndpoint() + "?query="
                                + query.getQueryString()};
    EXPECT_EQ(expected_query , obtained_query);
}


// Testing the PuppetDB connector

class ConnectionTest : public ::testing::Test {

  protected:

    CURL* curl_handle_;

    virtual void SetUp() {
        curl_handle_ = curl_easy_init();
    }

    virtual void TearDown() {
        curl_easy_cleanup(curl_handle_);
    }
};


TEST_F(ConnectionTest, ConnectWithoutSSL) {
    PuppetdbConnector connector {"spam", 42, ApiVersion::v2};

    EXPECT_FALSE(connector.isSecure());
}

TEST_F(ConnectionTest, ValidConnectionWithoutSSL) {
    PuppetdbConnector connector {"eggs", 42, ApiVersion::v3};

    EXPECT_TRUE(connector.isValid());
    EXPECT_EQ("", connector.getErrorMessage());
}

TEST_F(ConnectionTest, ConnectDefaultPortAndVersion) {
    PuppetdbConnector connector {"spam"};
    Query query("facts");
    std::string expected_url {"http://spam:8080/v3/facts"};

    EXPECT_FALSE(connector.isSecure());
    EXPECT_TRUE(connector.isValid());
    EXPECT_EQ(expected_url, connector.getQueryUrl(query, curl_handle_));
}

TEST_F(ConnectionTest, ConnectWithoutHost) {
    PuppetdbConnector connector {""};

    EXPECT_FALSE(connector.isValid()); // No hostname!
    EXPECT_EQ("No hostname was specified.", connector.getErrorMessage());
}

TEST_F(ConnectionTest, ConnectWithSSL) {
    PuppetdbConnector connector {"fake_host",
                                 "/fake/path/ca.cer",
                                 "/fake/path/host.cer",
                                 "/fake/path/host.key"};

    EXPECT_TRUE(connector.isSecure());
    EXPECT_FALSE(connector.isValid()); // Certificates are fake!
    EXPECT_EQ("Invalid certificate: /fake/path/ca.cer",
              connector.getErrorMessage());
}

TEST_F(ConnectionTest, performQueryWithInvalidConnection) {
    PuppetdbConnector connector {""};
    Query query {"spam"};
    std::string result = connector.performQuery(query);

    EXPECT_EQ("", result);
}

TEST_F(ConnectionTest, performQueryWithSimpleResult) {
    MockSetupConnector mock_connector {"bar"};
    Query query {"foo"};

    EXPECT_TRUE(mock_connector.isValid());

    EXPECT_CALL(mock_connector, setupAndPerform(testing::_))
        .WillOnce(testing::Return("simple_result"));

    std::string result = mock_connector.performQuery(query);

    EXPECT_EQ("simple_result", result);
    EXPECT_EQ("", mock_connector.getErrorMessage());
}

TEST_F(ConnectionTest, httpQuery) {
    MockURLConnector mock_connector {"bar"};
    Query query {"foo"};
    EXPECT_CALL(mock_connector, getQueryUrl(testing::_, testing::_))
        .WillOnce(testing::Return("http://example.com"));

    std::string result = mock_connector.performQuery(query);

    EXPECT_FALSE(result.size() == 0);
}

TEST_F(ConnectionTest, multipleHttpQueries) {
    MockURLConnector mock_connector {"spam"};
    Query first_query {"eggs"};
    Query second_query {"beans"};
    EXPECT_CALL(mock_connector, getQueryUrl(testing::_, testing::_))
        .WillOnce(testing::Return("http://example.com"))
        .WillOnce(testing::Return("http://goolge.com"));

    std::string first_result = mock_connector.performQuery(first_query);
    std::string second_result = mock_connector.performQuery(second_query);

    EXPECT_FALSE(first_result.size() == 0);
    EXPECT_FALSE(second_result.size() == 0);
}

#ifdef PUPET_AGENT_TEST
    TEST_F(ConnectionTest, testingPuppetSSL) {
        PuppetdbConnector connector {
            "master",
            "/etc/puppetlabs/puppet/ssl/certs/ca.pem",
            "/etc/puppetlabs/puppet/ssl/certs/centos65a.pem",
            "/etc/puppetlabs/puppet/ssl/private_keys/centos65a.pem"};
        Query query {"facts"};
        std::string result = connector.performQuery(query);
        std::cout << "Facter result:\n" << result << "\n\n";
        EXPECT_FALSE(result.size() == 0);
    }
#endif

}  // namespace PuppetdbQuery

#endif  // PUPPETDBQUERY_TEST_PUPPETDB_QUERY_TEST_H_


examples/example1.cpp
/*
    example1.cpp
    ============

    Basic usage of the libpuppetdb (without SSL)

    On OS X, compile with:
        c++ -std=c++11 -lcurl example1.cpp -o example

    By default, puppetdb only accepts http over 8080 as localhost.
    Check http://goo.gl/HGV4Mr.

    Run with:
        ./example hostname puppetdb_endpoint puppetdb_query
    For instance:
        ./exmple localhost facts ""
        ./example localhost nodes '["=", "name", "master"]'
    after running:
        ssh -NL 8080:localhost:8080 {user}@{puppetdb host} -f
*/


#include "../include/libpuppetdb/libpuppetdb.h"
#include <string>

using namespace LibPuppetdb;

int main(int argc, char* argv[]) {
    if (argc != 4)
        return 1;

    std::string hostname { argv[1] };
    std::string endpoint { argv[2] };
    std::string query_string { argv[3] };

    try {
        PuppetdbConnector connector { hostname, 8080, ApiVersion::v3 };

        Query query { endpoint, query_string };

        auto result = connector.performQuery(query);

        std::cout << "Result:\n" << result << "\n\n";
        std::cout << "Performed query: " << connector.getPerformedQueryUrl()
                  << "\n";
        return 0;
    } catch (connector_error& e) {
        std::cout << "Failed to initialize the connector: " << e.what() << "\n";
        return 2;
    } catch (query_error& e) {
        std::cout << "Failed to initialize the query: " << e.what() << "\n";
        return 3;
    } catch (processing_error& e) {
        std::cout << "Failed to perform the query: " << e.what() << "\n";
        return 4;
    }
}


include/libpuppetdb/libpuppetdb.h
#ifndef LIBPUPPETDB_INCLUDE_PUPPETDB_QUERY_H_
#define LIBPUPPETDB_INCLUDE_PUPPETDB_QUERY_H_

/*
 *                        libpuppetdb
 *
 * Simple library to execute Puppet DB queries in a synchronous.
 * To use it, create a PuppetdbConnector instance (SSL-enabled or not)
 * and call performQuery() on Query instances.
 *
 * The user can spcecify the puppetdb API version (v2, v3, or v4, as
 * for the ApiVersion enumeration) when instantiating the connector;
 * this will only affect the api URL since the query structure does
 * not change change across the supported versions. Also, the user is
 * responsible for specifying an endpoint compatible with the version.
 *
 * Multiple queries can be executed once a connection is set.
 * See example1.cpp and refer to README.
 *
 *
 * REQUIREMENTS
 *
 * - C++11 compiler
 * - libcurl, SSL-enabled for secure connections
 *
 *
 * ASSUMPTIONS
 *
 * - query format (query string is optional):
 *   {prot}://{hostname}:{port}/{version}/{endpoint}?query=<query_string>
 * - it is up to the client program to provide an endpoint compatible
 *   with the specified API version
 * - it is up to the client program to provide the query string in
 *   the puppetdb language and to process results, i.e.:
 *      - the interface expects the query string as a string argument
 *      - the interface returns the json result data as a string
 * - secure connection requires ca_cert, node_cert, and private_key
 *
 */

#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <fstream>
#include <stdexcept>

#include <curl/curl.h>

namespace LibPuppetdb {

//
// Tokens
//

// puppedbquery version
static const std::string VERSION_STRING { "0.2.0" };

// Protocol and ports
static const int PUPPETDB_HTTP_PORT { 8080 };
static const int PUPPETDB_SECURE_PORT { 8081 };

// PuppetDB api version
enum class ApiVersion { v2, v3, v4 };
static std::map<ApiVersion, const std::string> ApiVersionsMap {
    { ApiVersion::v2, "v2" },
    { ApiVersion::v3, "v3" },
    { ApiVersion::v4, "v4" }
};
static const ApiVersion API_VERSION_DEFAULT { ApiVersion::v4 };

//
// Errors
//

// Parent error class
class libpuppetdb_error : public std::runtime_error {
  public:
    explicit libpuppetdb_error(std::string const& msg) : std::runtime_error(msg) {}
};

// Invalid query
class query_error : public libpuppetdb_error {
  public:
    explicit query_error(std::string const& msg) : libpuppetdb_error(msg) {}
};

// Invalid connector
class connector_error : public libpuppetdb_error {
  public:
    explicit connector_error(std::string const& msg) : libpuppetdb_error(msg) {}
};

// Error due to a query processing failure
class processing_error : public libpuppetdb_error {
  public:
    explicit processing_error(std::string const& msg) : libpuppetdb_error(msg) {}
};

//
// Utility functions
//

inline bool fileExists (const std::string& file_path) {
    bool exists { false };
    if (!file_path.empty()) {
        std::ifstream f_s { file_path };
        exists = f_s.good();
        f_s.close();
    }
    return exists;
}

//
// Query
//

class Query {
  public:
    Query() = delete;

    /// The query string must be URL-encoded.
    /// Throws a query_error in case the endpoint is an empty string.
    Query(std::string endpoint, std::string query_string = "")
            : endpoint_ { endpoint },
              query_string_ { query_string } {
        if (endpoint.empty()) {
            throw query_error { "no endpoint specified" };
        }
    }

    std::string getEndpoint() {
        return endpoint_;
    }

    std::string getQueryString() {
        return query_string_;
    }

  private:
    // Endpoint
    std::string endpoint_;

    // Query string
    std::string query_string_;
};

//
// Query result
//

struct QueryResult {
    char* content;
    size_t size;

    static size_t callback(void* contents, size_t size, size_t nmemb,
                           void *userp) {
        static_cast<std::string*>(userp)->append(static_cast<char*>(contents),
                                                 size * nmemb);
        return size * nmemb;
    }
};

//
// PuppetdbConnector
//

class PuppetdbConnector {
  public:
    PuppetdbConnector() = delete;

    /// Constructor for HTTP connector
    /// Throws a connector_error in case if the hostname is an empty
    /// string
    PuppetdbConnector(const std::string& hostname,
                      const int port = PUPPETDB_HTTP_PORT,
                      const ApiVersion api_version = API_VERSION_DEFAULT)
            : hostname_ { hostname },
              port_ { port },
              api_version_ { api_version },
              is_secure_ { false },
              performed_query_url_ {} {
        checkHostname();
    }

    /// Constructor for SSL connector
    /// Throws a connector_error in case if the hostname is an empty
    /// string, if libcurl is not SSL enabled, or if any of the
    /// specified certificates is not an existent file
    PuppetdbConnector(const std::string& hostname,
                      const std::string& ca_crt_path,
                      const std::string& client_crt_path,
                      const std::string& client_key_path,
                      const int port = PUPPETDB_SECURE_PORT,
                      const ApiVersion api_version = API_VERSION_DEFAULT)
            : hostname_ { hostname },
              port_ { port },
              api_version_ { api_version },
              ca_crt_path_ { ca_crt_path },
              client_crt_path_ { client_crt_path },
              client_key_path_ { client_key_path },
              is_secure_ { true },
              performed_query_url_ {} {
        checkHostname();
        checkSSLSupport();
        checkCertificates();
    }

    /// Returns true if the Connector instance uses SSL, false
    /// otherwise
    bool isSecure() const {
        return is_secure_;
    }

    /// Returns the PuppetDB query result (json format) as a string
    /// Throws a processing_error in case of failure
    std::string performQuery(Query& query) {
        return setupAndPerform(query);
    }

    /// Returns the URL used to perform the PuppetDB query
    std::string getPerformedQueryUrl() const {
        return performed_query_url_;
    }

    // NB: this method is public and virtual for testing purposes
    virtual std::string getQueryUrl(Query& query, CURL* curl) {
        std::string protocol { isSecure() ? "https" : "http" };
        std::string endpoint_and_query = query.getEndpoint();
        std::string query_str = query.getQueryString();

        if (!query_str.empty()) {
            // URL encode
            char* encoded_query = curl_easy_escape(curl, query_str.c_str(), 0);

            if (encoded_query == nullptr) {
                throw processing_error { "failed to encode the query URL" };
            }

            endpoint_and_query += "?query=" + std::string { encoded_query };

            // Free the memory used by curl_easy_escape()
            curl_free(encoded_query);
        }

        return protocol + "://" + hostname_ + ":" + std::to_string(port_)
               + "/" + getApiVersion() + "/" + endpoint_and_query;
    }

  private:
    // PuppetDB host, port, and api version
    std::string hostname_;
    int port_;
    ApiVersion api_version_;

    // File paths used for the SSL connection
    std::string ca_crt_path_;
    std::string client_crt_path_;
    std::string client_key_path_;

    // SSL flag
    bool is_secure_;

    // The URL of the performed query
    std::string performed_query_url_;

    void checkHostname() {
        if (hostname_.empty()) {
            throw connector_error { "no hostname specified" };
        }
    }

    void checkSSLSupport() {
        curl_version_info_data* info_data = curl_version_info(CURLVERSION_NOW);
        if (!(info_data->features & CURL_VERSION_SSL)) {
            throw connector_error { "libcurl is not SSL enabled" };
        }
    }

    void checkCertificates() {
        std::vector<std::string> certificates { ca_crt_path_,
                                                client_crt_path_,
                                                client_key_path_ };
        for (auto cert : certificates) {
            if (cert.empty()) {
                throw connector_error { "not all certificates were specified" };
            }
        }

        for (auto cert : certificates) {
            if (!fileExists(cert)) {
                throw connector_error { "invalid certificate file: " + cert };
            }
        }
    }

    std::string getApiVersion() {
        return ApiVersionsMap[api_version_];
    }

    // NB: this is virtual to enable mocking
    virtual std::string setupAndPerform(Query& query) {
        std::string result_buffer {};

        // libcurl handle
        CURL* curl { curl_easy_init() };

        if(curl) {
            performed_query_url_ = getQueryUrl(query, curl);

            // Configure the libcurl handle
            curl_easy_setopt(curl, CURLOPT_URL,
                             performed_query_url_.c_str());
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION,
                             QueryResult::callback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &result_buffer);

            if (isSecure()) {
                setSSLOptions(curl);
            }

            // Perform the query
            CURLcode return_code { curl_easy_perform(curl) };

            if (return_code != CURLE_OK) {
                throw processing_error {
                    std::string(curl_easy_strerror(return_code)) };
            }

            // Close all libcurl connections
            curl_easy_cleanup(curl);
        }
        return result_buffer;
    }

    void setSSLOptions(CURL* curl) {
        curl_easy_setopt(curl, CURLOPT_SSLCERT, client_crt_path_.c_str());
        curl_easy_setopt(curl, CURLOPT_SSLKEY, client_key_path_.c_str());
        curl_easy_setopt(curl, CURLOPT_CAINFO, ca_crt_path_.c_str());
    }
};

}  // namespace LibPuppetdb

#endif  // LIBPUPPETDB_INCLUDE_PUPPETDB_QUERY_H_


include/puppetdbquery/puppetdbquery.h
#ifndef PUPPETDBQUERY_INCLUDE_PUPPETDB_QUERY_H_
#define PUPPETDBQUERY_INCLUDE_PUPPETDB_QUERY_H_


/*
 *                        libpuppetdb
 *
 * Simple library to execute Puppet DB queries in a synchronous.
 * To use it, create a PuppetdbConnector instance (SSL-enabled or not)
 * and call performQuery() on Query instances.
 * Multiple queries can be executed once a connection is set.
 * See example1.cpp and refer to README.
 *
 *
 * ASSUMPTIONS
 *
 * - C++11
 * - query format (query string is optional):
 *   {prot}://{hostname}:{port}/{version}/{endpoint}?query=<query_string>
 * - it is up to the client program to provide the query string in
 *   the puppetdb language and to process results, i.e.:
 *      - the interface expects the query string as a string argument;
 *      - the interface returns the json result data as a string.
 * - secure connection requires ca_cert, node_cert, and private_key.
 *
 */


#include <string>
#include <vector>
#include <map>
#include <curl/curl.h>
#include <iostream>
#include <fstream>

namespace PuppetdbQuery {

// puppedbquery version
static const std::string VERSION_STRING { "0.1.0" };

// Protocol and ports
static const int PUPPETDB_HTTP_PORT { 8080 };
static const int PUPPETDB_SECURE_PORT { 8081 };

// PuppetDB api version
enum class ApiVersion { v2, v3 };
std::map<ApiVersion, const std::string> ApiVersionsMap {
    { ApiVersion::v2, "v2" },
    { ApiVersion::v3, "v3" }
};
static const ApiVersion API_VERSION_DEFAULT { ApiVersion::v3 };

// Error codes (NB: starting at 100 to avoid clashing with curl ones)
enum class ErrorCode {
    OK = 100,
    INVALID_CONNECTION = 101,
    INVALID_QUERY = 102,
    URL_ENCODING_FAILURE = 103
};

// Utility function
inline bool fileExists (const std::string& file_path) {
    std::ifstream f_s { file_path };
    bool exists { f_s.good() };
    f_s.close();
    return exists;
}


// Query

class Query {

  public:

    /// The query string must be URL-encoded
    Query(std::string endpoint, std::string query_string = "")
        : endpoint_ { endpoint },
          query_string_ { query_string } {
        if (endpoint.empty()) {
            error_code_ = static_cast<int>(ErrorCode::INVALID_QUERY);
        } else {
            error_code_ = static_cast<int>(ErrorCode::OK);
        }
    }

    bool isValid() const {
        return error_code_ == static_cast<int>(ErrorCode::OK);
    }

    int getErrorCode() const {
        return error_code_;
    }

    void setErrorCode(int error_code) {
        error_code_ = error_code;
    }

    std::string getEndpoint() {
        return endpoint_;
    }

    std::string getQueryString() {
        return query_string_;
    }

  private:

    // Endpoint
    std::string endpoint_;

    // Query string
    std::string query_string_;

    // Keeps track of invalid initialization and query status
    int error_code_;
};


// Query result

struct QueryResult {
    char* content;
    size_t size;

    static size_t callback(void* contents, size_t size, size_t nmemb,
                           void *userp) {
        static_cast<std::string*>(userp)->append(static_cast<char*>(contents),
                                                 size * nmemb);
        return size * nmemb;
    }
};


// Connector

class PuppetdbConnector {

  public:

    PuppetdbConnector() = delete;

    /// Constructor for HTTP connector
    PuppetdbConnector(const std::string& hostname,
                      const int port = PUPPETDB_HTTP_PORT,
                      const ApiVersion api_version = API_VERSION_DEFAULT)
        : hostname_ { hostname },
          port_ { port },
          api_version_ { api_version },
          is_secure_ { false },
          is_valid_ { true } {
        checkHostname();
    }

    /// Constructor for SSL connector
    PuppetdbConnector(const std::string& hostname,
                      const std::string& ca_crt_path,
                      const std::string& client_crt_path,
                      const std::string& client_key_path,
                      const int port = PUPPETDB_SECURE_PORT,
                      const ApiVersion api_version = API_VERSION_DEFAULT)
        : hostname_ { hostname },
          port_ { port },
          api_version_ { api_version },
          ca_crt_path_ { ca_crt_path },
          client_crt_path_ { client_crt_path },
          client_key_path_ { client_key_path },
          is_secure_ { true },
          is_valid_ { true } {

        if (!checkHostname()) {
            return;
        }
        if (!checkSSLSupport()) {
            return;
        }
        checkCertificates();
    }

    bool isValid() const {
        return is_valid_;
    }

    bool isSecure() const {
        return is_secure_;
    }

    std::string getErrorMessage() const {
        return error_msg_;
    }

    /// Returns the PuppetDB query result (json format) as a string.
    /// Returns an empty string in case of an invalid query or
    /// connector.
    std::string performQuery(Query& query) {
        if (isValid()) {
            if (query.isValid()) {
                return setupAndPerform(query);
            }
        } else {
            query.setErrorCode(static_cast<int>(ErrorCode::INVALID_CONNECTION));
        }
        return "";
    }

    /// In case a query string is specified, the method returns an
    /// empty string if the encoding fails.
    virtual std::string getQueryUrl(Query& query, CURL* curl) {
        std::string protocol { isSecure() ? "https" : "http" };
        std::string endpoint_and_query = query.getEndpoint();
        std::string query_str = query.getQueryString();

        if (!query_str.empty()) {

            // URL encode
            char* encoded_query = curl_easy_escape(
                curl, ("query=" + query_str).c_str(), 0);

            if (encoded_query == nullptr) {
                return "";
            }
            endpoint_and_query += "?" + std::string { encoded_query };

            // Free the memory used by curl_easy_escape()
            curl_free(encoded_query);
        }

        return protocol + "://" + hostname_ + ":" + std::to_string(port_)
               + "/" + getApiVersion() + "/" + endpoint_and_query;
    }

  private:

    // PuppetDB host, port, and api version
    std::string hostname_;
    int port_;
    ApiVersion api_version_;

    // File paths used for the SSL connection
    std::string ca_crt_path_;
    std::string client_crt_path_;
    std::string client_key_path_;

    // SSL flag
    bool is_secure_;

    // This mimics RAII (no exception handling - flags failures)
    bool is_valid_;
    std::string error_msg_;

    bool checkHostname() {
        if (hostname_.empty()) {
            is_valid_ = false;
            error_msg_ = "No hostname was specified.";
            return false;
        }
        return true;
    }

    bool checkSSLSupport() {
        curl_version_info_data* info_data = curl_version_info(CURLVERSION_NOW);
        if (info_data->features & CURL_VERSION_SSL) {
            return true;
        } else {
            is_valid_ = false;
            error_msg_ = "libcurl is not SSL enabled.";
            return false;
        }
    }

    bool checkCertificates() {
        for (auto attr : std::vector<std::string> { hostname_, ca_crt_path_,
                client_crt_path_, client_key_path_ }) {
            if (attr.empty()) {
                is_valid_ = false;
                error_msg_ = "Not all certificates were specified.";
                return false;
            }
        }

        for (auto attr : std::vector<std::string> { ca_crt_path_,
                client_crt_path_, client_key_path_ }) {
            if (!fileExists(attr)) {
                is_valid_ = false;
                error_msg_ = "Invalid certificate: " + attr;
                return false;
            }
        }
        return true;
    }

    std::string getApiVersion() {
        return ApiVersionsMap[api_version_];
    }

    // NB: this is virtual to enable mocking
    virtual std::string setupAndPerform(Query& query) {

        std::string result_buffer {};

        // libcurl handle
        CURL* curl { curl_easy_init() };

        if(curl) {

            std::string url { getQueryUrl(query, curl) };

            if (url.empty()) {
                query.setErrorCode(static_cast<int>(
                    ErrorCode::URL_ENCODING_FAILURE));
            } else {

                // Configure the libcurl handle
                curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
                curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION,
                                 QueryResult::callback);
                curl_easy_setopt(curl, CURLOPT_WRITEDATA, &result_buffer);

                if (isSecure()) {
                    setSSLOptions(curl);
                }

                // Perform the query
                CURLcode return_code {curl_easy_perform(curl)};

                if (return_code != CURLE_OK) {
                    query.setErrorCode(static_cast<int>(return_code));
                }
            }
            // Close all libcurl connections
            curl_easy_cleanup(curl);
        }
        return result_buffer;
    }

    void setSSLOptions(CURL* curl) {
        curl_easy_setopt(curl, CURLOPT_SSLCERT, client_crt_path_.c_str());
        curl_easy_setopt(curl, CURLOPT_SSLKEY, client_key_path_.c_str());
        curl_easy_setopt(curl, CURLOPT_CAINFO, ca_crt_path_.c_str());
    }

};


}  // namespace PuppetdbQuery

#endif  // PUPPETDBQUERY_INCLUDE_PUPPETDB_QUERY_H_
